pkg/export/export.go

```diff
@@ -467,43 +482,72 @@ func (e *Exporter) Run(ctx context.Context) error {
 	// Starting index when iterating over shards. This ensures we don't always start at 0 so that
 	// some shards may never be sent in a busy collector.
 	shardOffset := 0
+	drainShards := func() bool {
+		sendIterations.Inc()
+		// Drain shards to fill up the batch.
+		//
+		// If the shard count is high given the overall throughput, a lot of shards may
+		// be packed into the same batch. A slow request will then block all those shards
+		// from further parallel sends.
+		// If this becomes a problem (especially when we grow maximum batch size), consider
+		// adding a heuristic to send partial batches in favor of limiting the number of
+		// shards they span.
+		i := 0
+		drained := true
+		for ; i < len(e.shards); i++ {
+			shardOffset = (shardOffset + 1) % len(e.shards)
+			shard := e.shards[shardOffset]
+
+			if took := shard.fill(batch); took > 0 {
+				drained = false
+				pendingShards = append(pendingShards, shard)
+			}
+			if batch.full() {
+				send()
+			}
+		}
+		// TODO(macxamin): May not be needed. Remove in a future release.
+		// If we didn't make a full pass over all shards, there may be more work.
+		if i < len(e.shards) {
+			e.triggerNext()
+		}
+		return drained
+	}
+
+	go func() {
+		// This routine monitors for a terminate signal. On termination, it will stop
+		// the exporter from recieving new data and try to drain the remaining shards
+		// within the CancelTimeout. This is done to prevent data loss during a shutdown.
+		<-parentCtx.Done()
+
+		e.mtx.Lock()
+		e.opts.Disable = true
+		e.mtx.Unlock()
+
+		done := make(chan struct{})
+		go func() {
+			drained := false
+			for !drained {
+				drained = drainShards()
+			}
+			close(done)
+		}()
+		select {
+		case <-done:
+			cancel()
+		case <-time.After(cancelTimeout):
+			level.Debug(e.logger).Log("msg", "Exporter timed out without exporting all of the pending buffered data.")
+			cancel()
+		}
+	}()
```

Comment:
The main loop below keeps executing until we completed the draining in this function.
That means that there are now two goroutines reading from the shards and sending data. That means that some state (e.g. `pendingShards`) is accessed concurrently without synchronization. But more importantly, we may send samples for the same stream in parallel, causing out-of-order writes w.r.t to the timestamps.
If you run/build with `go -race` the race detector is compiled it and should reliably trigger for such scenarios. When testing code like this, using the race detector is always a good idea.

So what we need to do is have the Run's main loop below stop/enter final state when at the first opportunity after the `parentCtx` is cancelled. So you can replace `case <-ctx.Done()` with `case <-parentCtx.Done()`.  
Then you can inline the contents of this function into that `case`. You don't need to watch for `ctx.Done()` anymore as there's no one else but "ourselves" as this function who could cancel it. And we do so (delayed)  after seeing `parentCtx` being canceled.

After that we're guaranteed to only execute `drainShards` once at a time.
Then you'll want to pass `drainShards` an explicit context so that for the regular sending you can pass in `ctx` and for the drain on shutdown you can pass in a new context instantiated via `context.WithTimeout(context.Background(), cancelTimeout)`.

I think after that things should be safe. The previous discussion might have steered things in a slightly different direction – but some interactions also only became clear to me when seeing the code evolve.